LFS-377: Improve filter performance

veronikaslc opened the pull request:
Created property index for "sling:resourceSuperType".

Commit:
LFS-377: Improve filter performance
Work in progress

sdumitriu committed:
LFS-377: Improve filter performance
Made the resourceSuperType property index work

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/content/oak%3Aindex/slingResourceSuperType.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "jcr:primaryType": "oak:QueryIndexDefinition",
+  "jcr:name:propertyNames": [
```
@ccmbioinfo/lfs-devs Something to keep in mind, this is how we can enforce a type other than String from the JSON.

sdumitriu committed:
LFS-377: Improve filter performance
Pagination queries should specify the right node type to be able to use the right indexes

sdumitriu committed:
LFS-377: Improve filter performance
Index the jcr:created property for faster sorting

IntegralProgrammer commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/PaginationServlet.java:
```diff
@@ -72,9 +72,13 @@ public void doGet(final SlingHttpServletRequest request, final SlingHttpServletR
+        // If we want this query to be fast, we need to use the exact nodetype requested.
+        // This isn't very pretty, but it's fast: just remove the 'homepage' part of the resource type.
+        final String nodeType =
+            request.getResource().getResourceType().replace('/', ':').replaceFirst("sHomepage$", "");
```
Would a more elegant solution be to specify a property like `childNodeType` in `dataentry.cnd` such that `lfs:FormsHomepage` has its `childNodeType` set to `lfs:Form`, `lfs:SubjectsHomepage` has its `childNodeType` set to `lfs:Subject`, and `lfs:QuestionnairesHomepage` has its `childNodeType` set to `lfs:Questionnaire`? @sdumitriu

sdumitriu replied:
Yes, that would be more elegant.

veronikaslc replied:
@sdumitriu Addressed in the last commit

Commit:
LFS-377: Improve filter performance
Added a property childNodeType in dataentry.cnd

veronikaslc commented:
@sdumitriu Addressed codereview comments

sashaandjic reviewed (approved):
tested

sdumitriu committed:
LFS-377: Improve filter performance
Index the Answer nodes as well

sdumitriu committed:
[misc] Debug logging

sdumitriu committed:
LFS-377: Improve filter performance
More property indexes

sdumitriu committed:
Revert "[misc] Debug logging"

This reverts commit 7c036692033c1541140c47ea300ada93c5774948.

sdumitriu committed:
[cleanup] Added missing EOF newlines
